crates/matrix-sdk-crypto/src/session_manager/group_sessions/share_strategy.rs

```diff
@@ -1122,6 +1150,187 @@ mod tests {
+                SessionRecipientCollectionError::SendingFromUnverifiedDevice
+            ))
+        );
+
+        // Finally, after we trust our own cross-signing keys, key sharing
+        // should succeed.
+        machine
+            .import_cross_signing_keys(CrossSigningKeyExport {
+                master_key: KeyDistributionTestData::MASTER_KEY_PRIVATE_EXPORT.to_owned().into(),
+                self_signing_key: KeyDistributionTestData::SELF_SIGNING_KEY_PRIVATE_EXPORT
+                    .to_owned()
+                    .into(),
+                user_signing_key: KeyDistributionTestData::USER_SIGNING_KEY_PRIVATE_EXPORT
+                    .to_owned()
+                    .into(),
+            })
+            .await
+            .unwrap();
+
+        let requests = machine
+            .share_room_key(
+                fake_room_id,
+                iter::once(KeyDistributionTestData::dan_id()),
+                encryption_settings.clone(),
+            )
+            .await
+            .unwrap();
+
+        // Dan has two devices, but only one is cross-signed, so there should
+        // only be one key share.
+        assert_eq!(requests.len(), 1);
+    }
+
+    #[async_test]
+    async fn test_share_identity_strategy_report_pinning_violation() {
+        // Test that identity-based key sharing gives an error when a verified
+        // user changes their identity.
+        let machine: OlmMachine = OlmMachine::new(
+            KeyDistributionTestData::me_id(),
+            KeyDistributionTestData::me_device_id(),
+        )
+        .await;
+
+        machine.bootstrap_cross_signing(false).await.unwrap();
+
+        let keys_query = IdentityChangeDataSet::key_query_with_identity_a();
+        machine.mark_request_as_sent(&TransactionId::new(), &keys_query).await.unwrap();
+
+        let keys_query = MaloIdentityChangeDataSet::initial_key_query();
+        machine.mark_request_as_sent(&TransactionId::new(), &keys_query).await.unwrap();
+
+        // Simulate pinning violation, in which case the key sharing should fail.
+        let keys_query = IdentityChangeDataSet::key_query_with_identity_b();
+        machine.mark_request_as_sent(&TransactionId::new(), &keys_query).await.unwrap();
+        machine
+            .get_identity(IdentityChangeDataSet::user_id(), None)
+            .await
+            .unwrap()
+            .unwrap()
+            .other()
+            .unwrap()
+            .mark_as_previously_verified()
+            .await
+            .unwrap();
+
+        let keys_query = MaloIdentityChangeDataSet::updated_key_query();
+        machine.mark_request_as_sent(&TransactionId::new(), &keys_query).await.unwrap();
+        machine
+            .get_identity(MaloIdentityChangeDataSet::user_id(), None)
+            .await
+            .unwrap()
+            .unwrap()
+            .other()
+            .unwrap()
+            .mark_as_previously_verified()
+            .await
+            .unwrap();
```

Comment:
we're changing our own identity and that of someone else?

Reply:
We're changing two other users' identities.  Our own user is `KeyDistributionTestData::me_id()`